ambry-router/src/main/java/com/github/ambry/router/GetBlobOperation.java

```diff
@@ -1658,6 +1774,127 @@ private void onInvalidRange(Exception exception) {
+      super();
+      /**
+       * If cache hit, extract all relevant metadata and save in appropriate variables.
+       * If the entry gets evicted later, we are still safe as we have saved all the metadata.
+       */
+      blobType = BlobType.MetadataBlob;
+      compositeBlobInfo = blobMetadata.getCompositeBlobInfo();
+      totalSize = compositeBlobInfo.getTotalSize();
+      chunkMetadataList = compositeBlobInfo.getChunkMetadataList();
+      blobInfo = blobMetadata.getBlobInfo();
+      lifeVersion = blobInfo.getLifeVersion();
+      serverBlobProperties = blobInfo.getBlobProperties();
+      chunkIndexToBuf = new ConcurrentHashMap<>();
+      chunkIndexToBufWaitingForRelease = new ConcurrentHashMap<>();
+      successfullyDeserialized = true;
+      logger.info(String.format("Metadata cache-hit for blobId = %s", blobId));
+    }
+
+    /**
+     * Return {@link MessageFormatFlags} to associate with cached first chunk.
+     * For a cached metadata chunk, we just need to get blobInfo to validate it.
+     * We don't have to fetch the entire metadata chunk again.
+     * @return {@link MessageFormatFlags#BlobInfo}
+     */
+    @Override
+    MessageFormatFlags getOperationFlag() {
+      return MessageFormatFlags.BlobInfo;
+    }
+
+    /**
+     * On an invalid cache entry, set a {@link RouterErrorCode} exception for first chunk, mark the chunk
+     * as completed, and set the chunk counters such that the operation will be completed.
+     * @param routerException Exception due to which cache entry was invalid.
+     */
+    protected void onInvalidCacheEntry(RouterException routerException) {
+      setChunkException(routerException);
+      chunkCompleted = true;
+      retainChunkExceptionOnSuccess = true;
+      chunkIdIterator = null;
+      dataChunks = null;
+      numChunksTotal = 0;
+      numChunksRetrieved.set(0);
+    }
+
+    /**
+     * Validate cached metadata chunk and proceed to request data chunks
+     * @param payload Bytes received from backend
+     * @param messageMetadata  Metadata associated with messages sent out
+     * @param messageInfo Message info class that contains basic info about a message
+     * @throws IOException
+     * @throws MessageFormatException
+     */
+    @Override
+    void handleBody(InputStream payload, MessageMetadata messageMetadata, MessageInfo messageInfo)
+        throws IOException, MessageFormatException {
+      // If we find blob metadata in frontend cache, validate it and use it.
+      BlobProperties receivedBlobProperties = MessageFormatRecord.deserializeBlobProperties(payload);
+      /**
+       * Some cached variables are mutable.
+       * Update them and then compare if other immutable variables are same or not.
+       */
+      updateTtlIfRequired(serverBlobProperties, messageInfo);
+      blobInfo.setLifeVersion(lifeVersion);
+      lifeVersion = messageInfo.getLifeVersion();
+      if (!serverBlobProperties.equals(receivedBlobProperties)) {
```

Comment:
we have to override BlobProperties's equals method.

Reply:
yes done